EIPS/eip-2539.md

```diff
@@ -0,0 +1,266 @@
+- BLS12_377_PAIRING - to perform a pairing operations between a set of *pairs* of (G1, G2) points
+
+Multiexponentiation operation is included to efficiently aggregate public keys or individual signer's signatures during BLS signature verification.
+
+### Proposed addresses table
+
+|Precompile   |Address   |
+|---|---|
+|BLS12_377_G1ADD          | 0x13  |
+|BLS12_377_G1MUL          | 0x14  |
+|BLS12_377_G1MULTIEXP     | 0x15  |
+|BLS12_377_G2ADD          | 0x16  |
+|BLS12_377_G2MUL          | 0x17  |
+|BLS12_377_G2MULTIEXP     | 0x18  |
+|BLS12_377_PAIRING        | 0x19  |
+
+## Motivation
+Motivation of this precompile is to add a cryptographic primitive that allows to get 120+ bits of security for operations over pairing friendly curve compared to the existing BN254 precompile that only provides 80 bits of security. In addition it allows efficient one-time recursive proof aggregations, e.g. proofs about existence of BLS12-377 based signature.
+
+## Specification
+
+Curve parameters:
+
+BLS12-377 curve is fully defined by the following set of parameters (coefficient `A=0` for all BLS12 curves):
+
+```
+Base field modulus = 0x01ae3a4617c510eac63b05c06ca1493b1a22d9f300f5138f1ef3622fba094800170b5d44300000008508c00000000001
+B coefficient = 0x000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001
+Main subgroup order = 0x12ab655e9a2ca55660b44d1e5c37b00159aa76fed00000010a11800000000001
+Extension tower:
+Fp2 construction:
+Fp quadratic non-residue = 0x01ae3a4617c510eac63b05c06ca1493b1a22d9f300f5138f1ef3622fba094800170b5d44300000008508bffffffffffc
+Fp6/Fp12 construction:
+Fp2 cubic non-residue c0 = 0x000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
+Fp2 cubic non-residue c1 = 0x000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001
+Twist parameters:
+Twist type: D
+B coefficient for twist c0 = 0x000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
+B coefficient for twist c1 = 0x010222f6db0fd6f343bd03737460c589dc7b4f91cd5fd889129207b63c6bf8000dd39e5c1ccccccd1c9ed9999999999a
+Generators:
+G1:
+X = 0x008848defe740a67c8fc6225bf87ff5485951e2caa9d41bb188282c8bd37cb5cd5481512ffcd394eeab9b16eb21be9ef
+Y = 0x01914a69c5102eff1f674f5d30afeec4bd7fb348ca3e52d96d182ad44fb82305c2fe3d3634a9591afd82de55559c8ea6
+G2:
+X c0 = 0x018480be71c785fec89630a2a3841d01c565f071203e50317ea501f557db6b9b71889f52bb53540274e3e48f7c005196
+X c1 = 0x00ea6040e700403170dc5a51b1b140d5532777ee6651cecbe7223ece0799c9de5cf89984bff76fe6b26bfefa6ea16afe
+Y c0 = 0x00690d665d446f7bd960736bcbb2efb4de03ed7274b49a58e458c282f832d204f2cf88886d8c7c2ef094094409fd4ddf
+Y c1 = 0x00f8169fd28355189e549da3151a70aa61ef11ac3d591bf12463b01acee304c24279b83f5e52270bd9a1cdd185eb8f93
+Pairing parameters:
+|x| (miller loop scalar) = 0x8508c00000000001
+x is negative = false
+```
+
+#### Fine points and encoding of base elements
```

Comment:
```suggestion
### Fine points and encoding of base elements
```
You skipped `###` here, this can sometimes break some rendering systems particularly when they try to render table of contents.

Reply:
Same thing with all of the following sections, they are all 1 too deep.